add-bookings 0.1

Lamchi01 opened the pull request:
- Добавление функционала бронирования с проверками пользователя и владельца вещей
- Реализация хранения через Hibernate

Lamchi01 committed:
add-bookings 0.1
- Добавление функционала бронирования с проверками пользователя и владельца вещей
- Реализация хранения через Hibernate

Lamchi01 committed:
add-bookings 0.1
- Добавление функционала бронирования с проверками пользователя и владельца вещей
- Реализация хранения через Hibernate

Lamchi01 committed:
add-bookings 0.1
- Добавление функционала бронирования с проверками пользователя и владельца вещей
- Реализация хранения через Hibernate

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,50 @@
-}
\ No newline at end of file
+    private static final String HEADER_USER_ID = "X-Sharer-User-Id";
+    private final BookingServiceImpl bookingService;
```
Внедряй зависимости через интерфейсы, чтобы не привязываться к конкретной реализации

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/dto/CreateBookingDto.java:
```diff
@@ -0,0 +1,23 @@
+    @NotNull
+    private Long itemId;
+    @NotNull
+    @Future
```
Может быть текущее время, поэтому лучше использовать аннотацию `@FutureOrPresent`

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,20 +1,35 @@
+    @Column(name = "end_date", nullable = false)
+    private LocalDateTime end;
+    @OneToOne
+    @JoinColumn(name = "item_id", nullable = false)
     private Item item;
-    private Long booker;
+    @OneToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User booker;
```
Почему OneToOne? У одного предмета может быть много бронирований, как и у пользователя может быть создано много бронирований

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,123 @@
+                .orElseThrow(() -> new NotFoundException("Пользователь не найден"));
+        Item item = itemStorage.findById(createBookingDto.getItemId())
+                        .orElseThrow(() -> new NotFoundException("Предмет не найден"));
+        if (createBookingDto.getStart().equals(createBookingDto.getEnd())) {
+            throw new ValidationException("Дата начала бронирования не может быть равна дате окончания");
```
Это можно проверить и в контроллере

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,123 @@
+        if (createBookingDto.getStart().equals(createBookingDto.getEnd())) {
+            throw new ValidationException("Дата начала бронирования не может быть равна дате окончания");
+        }
+        if (!item.getAvailable()) {
+            throw new ValidationException("Предмет недоступен");
+        }
+        Booking booking = BookingMapper.toBookingNew(createBookingDto, item, user);
+        return BookingMapper.toBookingDto(bookingStorage.save(booking));
```
Нужно еще проверить, чтобы не было созданных бронирований в статусе APPROVED, которые пересекаются по времени с новым

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/exceptions/ErrorHandler.java:
```diff
@@ -20,13 +21,12 @@ public ErrorResponse handleBadRequest(ValidationException e) {
     }
 
-    @ExceptionHandler
+    @ExceptionHandler({EmailValidException.class, DataIntegrityViolationException.class})
```
Если используются абсолютно разные исключения, которые возникают при разных обстоятельствах, тогда лучше писать отдельные обработчики

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/dto/CommentDto.java:
```diff
@@ -0,0 +1,20 @@
+@ToString
+public class CommentDto {
+    private Long id;
+    @NotNull
+    private String text;
```
Для создания лучше использовать отдельную дтошку, обязателен только текст комментария. А эту дтошку использовать для ответов.

Плюс, текст не должен быть пустым. А тут пропустится пустой комментарий

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,55 +3,119 @@
+        List<Booking> bookings = bookingStorage.findAllByItemIdAndBookerId(itemId, authorId);
+        for (Booking booking : bookings) {
+            if (booking.getStatus() == Status.APPROVED && booking.getEnd().isBefore(LocalDateTime.now())) {
+                Comment comment = new Comment();
+                comment.setText(text.trim());
+                comment.setAuthor(author);
+                comment.setItem(item);
+                comment.setCreated(LocalDateTime.now());
```
Вынеси в маппер

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,14 +1,49 @@
+    @Transient
+    private List<Comment> comments = new ArrayList<>();
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Item item = (Item) o;
+        return Objects.equals(id, item.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(id);
+    }
```
А для чего equals() и hashCode()? Обычно для сущностей они не переопределяются

Lamchi01 committed:
add-bookings 0.2
Некоторые исправления
Чуть упрощение кода в методе поиска, безотложная загрузка зависимостей в сущностях

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/dto/CreateBookingDto.java:
```diff
@@ -15,9 +15,9 @@ public class CreateBookingDto {
     private LocalDateTime start;
     @NotNull
-    @Future
+    @FutureOrPresent
```
Тут лучше только `@Future`, я имел в виду только время старта брони, оно может совпадать с текущим

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -36,12 +36,14 @@ public BookingDto addBooking(CreateBookingDto createBookingDto, long userId) {
             throw new ValidationException("Предмет недоступен");
         }
+        List<Booking> bookings = bookingStorage.findAllByIntersectingStartAndEnd(
+                item.getId(), Status.APPROVED, Status.WAITING, createBookingDto.getStart(), createBookingDto.getEnd());
```
Смотри только по статусам `APPROVED`, неподтвержденные бронирования нас не интересуют
Статус можно указать в самом запросе, чтобы уменьшить количество передаваемых параметров

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/storage/BookingStorage.java:
```diff
@@ -0,0 +1,40 @@
+
+    // Поиск бронирований на пересечение дат
+    @Query("select b from Booking b " +
+            "where ?1 = b.item.id and ?2 = b.status and ?3 = b.status and ?4 <= b.start and ?5 >= b.end")
```
Условие по времени не совсем корректное, проверяется только полное перекрытие интервала, но может пересекаться еще частично

 Самое простое - это написать условие для не пересекающихся интервалов.
     Непересечение, это когда начало одного интервала больше конца другого интервала, или конец одного интервала меньше начала другого:
              a.start > b.end OR a.end < b.start
    Соответственно, пересечение - это отрицание того, что мы получили - т.е. если для интервалов выполняется:
             NOT( a.start > b.end OR a.end < b.start )
    Теперь раскрываем по правилам логики скобки и получаем a.start <= b.end AND a.end >= b.start
останется превратить это псевдоусловие в логическое условие на языке java

Точно такой же алгоритм уже встречался при реализации отсортированного списка в 7 или 8 спринте

Lamchi01 committed:
add-bookings 0.3
Поправил метод для поиска пересечений, немного уменьшил кода
